Add basic Personalization Tags support in the Gutenberg Email Editor [MAILPOET-6328]

lysyjan opened the pull request:
## Description

This PR adds the possibility of registering Personalization Tags implemented via HTML comments. Registered tags need to specify the name, tag, category, callback, and attributes.

## Code review notes

1. All new classes in the email editor package are covered with tests, but the implementation in the MailPoet plugin no. I want to focus on tests in the plugin in one of the follow-up tickets when we add more tags.
2. Replacement in attributes will be covered in one of the next PRs.
3. Insertion via UI will be added later.

## QA notes

- QA is not needed here or is optional by the reviewer.
- For testing, the Gutenberg plugin needs to be activated with the build from the latest trunk.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-6328]

## After-merge notes

_N/A_

## Tasks

- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6328]: https://mailpoet.atlassian.net/browse/MAILPOET-6328?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:email-editor/tags-api)

_The latest successful build from `email-editor/tags-api` will be used. If none is available, the link won't work._

lysyjan committed:
Add personalization-tag-registry class

The new class which allows register Personalization Tags and then read them.
[MAILPOET-6328]

lysyjan committed:
Add new class extending WP_HTML_Tag_Processor

The new class allows replacing tokens which is necessary for Personalization Tags implementation.
[MAILPOET-6328]

lysyjan committed:
Add new class Personalizer managing tags replacement

[MAILPOET-6328]

lysyjan committed:
Add Personalization Tags support for title

[MAILPOET-6328]

lysyjan committed:
Register subscriber Personalization Tags in gutenberg editor

This commit adds a new class which register MailPoet subscribers tags for the new email editor.
[MAILPOET-6328]

lysyjan committed:
Personalize email created via Gutenberg email editor

[MAILPOET-6328]

pavel-mailpoet commented on packages/php/email-editor/src/Engine/PersonalizationTags/class-personalization-tags-registry.php:
```diff
@@ -0,0 +1,75 @@
+	/**
+	 * Register a new personalization tag.
+	 *
+	 * @param string   $name        Unique identifier for the callback.
```
"Unique identifier for the callback"

Should this be the identifier of the tag? I think this needs a better description because I am not sure what is the difference between `$name`, `$tag`, and why do we need category here and what it does.

lysyjan replied:
You are right. The name should be displayed to users in the modal windows with tags overview or other UI parts. I will update the comment here.

pavel-mailpoet commented on packages/php/email-editor/src/Engine/PersonalizationTags/class-personalization-tags-registry.php:
```diff
@@ -0,0 +1,75 @@
+	 * @param string   $name        Unique identifier for the callback.
+	 * @param string   $tag         The tag to be used in the email content.
+	 * @param string   $category    The category of the personalization tag.
+	 * @param callable $callback    The callable function/method.
```
"The callable function/method." - this is technically correct, but as a reader, I have no idea what this is used for and when the callback is called.

lysyjan replied:
👍 I will update the comment. This is a callback that returns a value of the tag.

pavel-mailpoet commented on packages/php/email-editor/src/Engine/PersonalizationTags/class-personalization-tags-registry.php:
```diff
@@ -0,0 +1,75 @@
+	 */
+	public function register( string $name, string $tag, string $category, callable $callback, array $attributes = array() ): void {
+		if ( isset( $this->tags[ $tag ] ) ) {
+			return;
```
Is the silent return here the best approach? I haven't seen it used yet, but bugs like this are quite hard to debug. Should this throw an exception if it is not expected to call with an already registered tag. Since this is part of MailPoet would it be useful to at least log something to the error log?

lysyjan replied:
I considered using error_log, but it needs adding CodeSniffer ignore, which I wanted to avoid initially.
I decided to use the silent return in the first phase because if you call registration twice it could generate errors, but maybe it's correct 🤔

pavel-mailpoet replied:
I can see many approaches:

- silent return
- silently overwrite the original value
- throw an exception
- silent return with an `error_log`
- silent return with a MailPoet log

I am not saying one is the correct one, but if we want to go with one, it should probably be a conscious decision, and we should have a reason why.

pavel-mailpoet commented on packages/php/email-editor/src/Engine/PersonalizationTags/class-personalization-tags-registry.php:
```diff
@@ -0,0 +1,75 @@
+	}
+
+	/**
+	 * Retrieve a personalization tag by its tag.
```
The naming here is quite weird, and the comment doesn't help. Would an example help to understand it better?

lysyjan replied:
I was thinking about a name `token` instead of the tag. It sounds to me more descriptive. What do you think about it? I will try to improve the comments across all new files.

pavel-mailpoet replied:
Sounds good, but maybe also an example could help.

pavel-mailpoet commented on packages/php/email-editor/src/Engine/PersonalizationTags/class-personalization-tags-registry.php:
```diff
@@ -0,0 +1,75 @@
+	 * Retrieve a personalization tag by its tag.
+	 *
+	 * @param string $tag The tag of the personalization tag.
+	 * @return array|null The array data or null if not found.
```
Should we describe the array better? When you are outside of this class, having only an array as the description is difficult to work with and tooling doesn't help. Should we use a class instead of array? Or at least document the array [better](https://phpstan.org/writing-php-code/phpdoc-types#array-shapes)?

lysyjan replied:
👍 I will update the annotation.

pavel-mailpoet commented on packages/php/email-editor/src/Engine/class-email-editor.php:
```diff
@@ -61,30 +62,40 @@ class Email_Editor {
+	 * @param Patterns                      $patterns Patterns.
+	 * @param Settings_Controller           $settings_controller Settings controller.
+	 * @param Send_Preview_Email            $send_preview_email Preview email controller.
+	 * @param Personalization_Tags_Registry $personalization_tags_controller Preview email controller.
```
I don't think the comment here is correct.

pavel-mailpoet commented on packages/php/email-editor/src/Engine/PersonalizationTags/class-html-tag-processor.php:
```diff
@@ -0,0 +1,51 @@
+ */
+class HTML_Tag_Processor extends WP_HTML_Tag_Processor {
+	/**
+	 * List of deferred updates.
```
This comment is technically true. However, it lacks clarity. Would it be better to describe what those deferred updates are, why we need them, or how they work?

pavel-mailpoet commented on packages/php/email-editor/tests/integration/Engine/PersonalizationTags/HTML_Tag_Processor_Test.php:
```diff
@@ -0,0 +1,102 @@
+		$processor->replace_token( 'John!' );
+
+		// Verify deferred updates.
+		$deferred_updates = $this->getPrivateProperty( $processor, 'deferred_updates' );
```
I am not sure if having this test testing the internals of the class is the best approach. Do you think it would be better to test the outside behaviour of the class? How does the class behave from the outside if it does what we need from it?

lysyjan replied:
That sounds good. I will update the test.

pavel-mailpoet commented on packages/php/email-editor/src/Engine/PersonalizationTags/class-personalization-tags-registry.php:
```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * This file is part of the MailPoet plugin.
```
I am not sure I understand why is this class (and other classes from this Pull Request) part of the MailPoet plugin and not part of the Email Editor. I don't see anything MailPoet specific here.  We will need personalisation tags even in other integrations, won't we?

lysyjan replied:
You are right. I will update those comments. Or do you have another idea of what can be used here in the class header?

pavel-mailpoet replied:
Well, the comment is fine with me, but I don't understand why the files are in the MailPoet plugin and not in the core editor. They seem generic, and we will need them (or something similar) when we port the editor to Woo Core or Fruita, won't we?

pavel-mailpoet commented on packages/php/email-editor/src/Engine/PersonalizationTags/class-personalizer.php:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	/**
+	 * Set the context for the personalization.
```
Could you provide more details here on what the "context" is?

lysyjan replied:
Yes. I will add it.

pavel-mailpoet commented on packages/php/email-editor/src/Engine/PersonalizationTags/class-personalizer.php:
```diff
@@ -0,0 +1,102 @@
+
+declare(strict_types = 1);
+
+namespace MailPoet\EmailEditor\Engine\PersonalizationTags;
```
So this is the entry point to all this functionality, isn't it? Would it be better to have this class in this folder, and the "helper" classes like `HTML_Tag_Processor` and the Registry in a subfolder?

lysyjan replied:
Good idea. I had it in my mind, but for some reason, I didn't realize it.

pavel-mailpoet commented on mailpoet/lib/EmailEditor/Integrations/MailPoet/PersonalizationTagManager.php:
```diff
@@ -0,0 +1,42 @@
+
+  public function initialize() {
+    add_filter('mailpoet_email_editor_register_personalization_tags', function( Personalization_Tags_Registry $registry ): Personalization_Tags_Registry {
+      $registry->register(
```
I am looking at the signature/usage of this method and I am wondering if it would be better to use a class as the argument here and pass it in. Instead of having a method with five arguments we would have a single argument. What do you think about it?

lysyjan replied:
I am not sure how you mean it. Can you please write a part of the code here for better understanding?

pavel-mailpoet replied:
So the method `register` takes arguments "string $name, string $tag, string $category, callable $callback, array $attributes = array()". 
I am suggesting to create a class:

```
class PersonalizationTag {
	private string $name; 
	private string $tag; 
	private string $category; 
	private callable $callback; 
	private array $attributes = array();
  ...
}
```

and then use this class not just as an argument of the method `public function register( PersonalizationTag $tag )` but also inside the registry class instead of the array.

pavel-mailpoet reviewed (changes requested):
Thank you for the pull request. Overall, this looks great. I have some suggestions, and I am open to discussing them. I went commit by commit, so I saw some classes before their usage, which may have influenced some of my comments.

lysyjan committed:
Move Personalizer to a different namespace

[MAILPOET-6328]

lysyjan committed:
Add new class Personalization_Tag

The new class contains all necessary data to replace token with personalized value.
[MAILPOET-6328]

lysyjan committed:
Improve comments across Personalization Tags classes

[MAILPOET-6328]

lysyjan committed:
Refactor HTML_Tag_Processor_Test integration test

[MAILPOET-6328]

lysyjan commented:
@pavel-mailpoet Thanks for the great review. I hope I addressed all your suggestions.
I created a new Jira ticket related to the logs.

pavel-mailpoet commented on packages/php/email-editor/src/Engine/PersonalizationTags/class-personalization-tags-registry.php:
```diff
@@ -23,6 +23,7 @@ class Personalization_Tags_Registry {
 
 	/**
 	 * Initialize the personalization tags registry.
+	 * This method should be called only once.
```
Not for this Pull request, but should we somehow ensure that the function is called only once? After it is called, set some property and then prevent calling it again.
